ampd/src/stellar/rpc_client.rs

```diff
@@ -27,21 +27,58 @@ pub struct TxResponse {
 }
 
 const STATUS_SUCCESS: &str = "SUCCESS";
+const EXPECTED_OPERATION_COUNT_V4: usize = 1;
 
 impl From<(Hash, GetTransactionResponse)> for TxResponse {
     fn from((transaction_hash, response): (Hash, GetTransactionResponse)) -> Self {
-        // Protocol 23 (CAP-0067): Extract contract events from the unified events structure
-        // contract_events is Vec<Vec<ContractEvent>> (per operation), so we flatten it
-        let contract_events: Vec<ContractEvent> = response
-            .events
-            .contract_events
-            .into_iter()
-            .flatten()
-            .collect();
+        let transaction_hash = transaction_hash.to_string();
+
+        let (contract_events, successful) = match response.result_meta.as_ref() {
+            Some(TransactionMeta::V4(_)) => {
+                // Protocol 23: Soroban contract calls can only occur as a single operation in the tx
+                // Therefore, we extract the first (and only) operation, otherwise fail.
+                let contract_events = response.events.contract_events;
+                let op_count = contract_events.len();
+                let events = contract_events.into_iter().flatten().collect();
+
+                if op_count != EXPECTED_OPERATION_COUNT_V4 {
+                    warn!(
+                        tx_hash = %transaction_hash,
+                        operation_count = op_count,
+                        expected_count = EXPECTED_OPERATION_COUNT_V4,
+                        "V4 transaction operation count does not match expected single operation for Soroban contract call"
+                    );
+                }
+
+                (
+                    events,
+                    response.status == STATUS_SUCCESS && op_count == EXPECTED_OPERATION_COUNT_V4,
+                )
+            }
+            Some(TransactionMeta::V3(data)) => {
+                let events = data
+                    .soroban_meta
+                    .as_ref()
+                    .map(|meta| meta.events.to_vec())
+                    .unwrap_or_default();
+
+                (
+                    events.clone(),
+                    response.status == STATUS_SUCCESS && !events.is_empty(),
+                )
+            }
+            _ => {
+                warn!(
+                    tx_hash = %transaction_hash,
+                    "Unsupported or missing transaction metadata version"
+                );
+                (vec![], false)
+            }
+        };
 
         Self {
-            transaction_hash: transaction_hash.to_string(),
-            successful: response.status == STATUS_SUCCESS,
+            transaction_hash,
+            successful,
```

Comment:
the meaning of `successful` variable is being changed here. It doesn't represent whether the tx was successful anymore. Change the structure of the code instead to do this in a more clear way. E.g. Implement `TryFrom` instead of `From` trait, so if there's an error (e.g. invalid op count), we return an Err instead of just making things sort of work with existing types

Reply:
It feels weird to add logging in a From conversion util. Let the caller log the returned error instead